db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/NumericTypeCompiler.java

```diff
@@ -364,6 +364,57 @@ else if (typeId.isBigIntTypeId() &&
 			}
 			else if (typeId.isDecimalTypeId())
 			{
+				/* DB-9425
+				 * Make sure we still have enough digits for the integral part.
+				 * Reduce scale when necessary, or leave it when not possible.
+				 * Rules follow SQL Server 2019:
+				 * https://docs.microsoft.com/en-us/sql/t-sql/data-types/precision-scale-and-length-transact-sql?view=sql-server-ver15
+				 */
+				{
+					int lprec = leftType.getPrecision();
+					int lscale = leftType.getScale();
+					int rprec = rightType.getPrecision();
+					int rscale = rightType.getScale();
+					int integralNumDigits;
+
+					switch (operator) {
+						case TypeCompiler.PLUS_OP:
+						case TypeCompiler.MINUS_OP:
+							boolean addOne = (lprec != maxPrecision && rprec != maxPrecision);
+							integralNumDigits = Math.max(lprec - lscale, rprec - rscale) + (addOne ? 1 : 0);
+							if (integralNumDigits <= precision && integralNumDigits + scale > precision) {
+								scale = precision - integralNumDigits;
+							}
+							break;
+						case TypeCompiler.TIMES_OP: {
+							integralNumDigits = lprec + rprec + 1 - lscale - rscale;
+							/* In SQL Server, MIN_DECIMAL_DIVIDE_SCALE = 6. In splice, it's 4.
+							 * We have to use 6 in multiplication case, otherwise TPC-H results would be
+							 * off too much because we round intermediate results too early.
+							 */
+							if (integralNumDigits < maxPrecision - MIN_DECIMAL_MULTIPLICATION_SCALE) {
+								scale = Math.min(scale, maxPrecision - integralNumDigits);
+							}
+							else if (scale > MIN_DECIMAL_MULTIPLICATION_SCALE) {
```

Comment:
Now that you introduced MIN_DECIMAL_MULTIPLICATION_SCALE, would there be an impact on the below code which looks at MIN_DECIMAL_DIVIDE_SCALE only. Could you double check the effect of your change on it:
```
				// Use high precision for decimals, so we don't overflow.
				// But don't use max precision to leave room for 4 digits to
				// the right of the decimal place for operations like AVG.
				// For the old client, use the old size of 31 for consistent behavior.
				int scaleAllowance = Math.max(0, MIN_DECIMAL_DIVIDE_SCALE - Math.max(scale, 0));
				int maxImplicitCastPrecision = precision > OLD_MAX_DECIMAL_PRECISION_SCALE ||
				                               supportsDecimal38()  ?
				      MAX_DECIMAL_PRECISION_WITH_RESERVE_FOR_SCALE + scaleAllowance :
				      OLD_MAX_DECIMAL_PRECISION_SCALE;
				if (precision < maxImplicitCastPrecision)
				    precision = maxImplicitCastPrecision;
```
However, I have some reservation about this code to use the highest precision possible for an operation on decimals(See DB-9639).

Reply:
Right, this part basically promotes the precision to at least 34. If the determined `scale` is smaller than 4, then there would be some digits allowance for the integral part, thus total precision is further increased.

`MIN_DECIMAL_MULTIPLICATION_SCALE` should have no impact here. If the initial scale was smaller than 6, say 3, then there will be no scale reduction in case of an multiplication. This part sees that there is 4-3=1 digit allowance for the integral part, and promote the type to (35,3).

If for any reason we reduce the scale to 6, then there will be no scale allowance. So if the initial precision was smaller than 34, final type would be (34,6), otherwise it's (precision,6).

Key idea is, when we reach this part, we know that we either have enough digits for integral part (scale reduced) or we cannot prevent overflow (scale untouched). In both cases, increasing precision do not affect correctness.

BUT!!!!!

This logic seems very strange to me. In my opinion, we should not blindly increase precision but to deduce it properly. If (20,8) is enough, then we should not increase it to (34,8).